pkg/ovs/openflow/ofctrl_builder.go

```diff
@@ -40,10 +41,36 @@ func (b *ofFlowBuilder) MatchTunMetadata(index int, data uint32) FlowBuilder {
 	return b
 }
 
-// MatchVLAN matches the VLAN VID. It holds the VLAN VID in its least significant 12 bits.
+// MatchVLAN can be used as follows:
+// - to match the packets of a specific VLAN, there are two cases:
+//   - to match VLAN 0, nonVLAN should be false, vlanID should be 0, value of vlanMask must be 0x1fff.
+//   - VLAN 1-4095, nonVLAN should be false, vlanID should be the VLAN ID, vlanMask can be nil, or its value can be 0x1fff.
+//
+// - to match the packets of all VLANs, nonVLAN should be false, vlanID must be 0, value of vlanMask must be 0x1000.
+// - to match the packets of non-VLAN, nonVLAN should be true, vlanID must be 0, vlanMask can be nil, or its value can be 0x1000.
 func (b *ofFlowBuilder) MatchVLAN(nonVLAN bool, vlanID uint16, vlanMask *uint16) FlowBuilder {
-	// TODO(gran): correct matchStr
-	matchStr := fmt.Sprintf("dl_vlan=%d", vlanID)
+	var vlanMaskValue uint16
+	if vlanMask == nil {
+		// To match the packets of a VLAN whose VLAN ID is not 0, when vlanMask is nil, set the value of vlanMask to 0x1ffff.
+		if vlanID != 0 {
+			vlanMaskValue = uint16(openflow15.OFPVID_PRESENT | protocol.VID_MASK)
+		}
+		// To match the packets of non-VLAN, when vlanMask is nil, set the value of vlanMask to 0x1000.
+		if nonVLAN {
+			vlanMaskValue = uint16(openflow15.OFPVID_PRESENT)
+		}
+		vlanMask = &vlanMaskValue
+	}
+
+	value := vlanID | openflow15.OFPVID_PRESENT
```

Comment:
this seems wrong. If I want to match non vlan, I set vlanID to 0, `value` will be `openflow15.OFPVID_PRESENT`, then matchStr would get `vlan_tci=0x1000/0x1000` regardless. If the comment is correct, it's worth to add an unit test for it as it's error prone.

Reply:
Yes, I'll correct it.